NET-2397: Add readme.md to upgrade test subdirectory (#16610)
hashicorp/consul: test/integration/consul-container/test/README.md

Reviewer: Should this file actually be in `test/upgrade`?

# Upgrade Integration Tests

> The upgrade integration test is responsible for ensuring that users are able to upgrade between consul versions without experiencing difficulties. The upgrade tests does the following:
> - Ensure that customers are able to upgrade to a new Consul version, and new and existing features work as expected. 
> - This tests aims to ensure a high degree of interoperability between versions of Consul by supporting `n-2` and `n-1` upgrades, ensuring that upgrading to newer versions does not degrade or break functionalities.
>     - Validate users are able to perform a direct upgrade of Consul. For example upgrading from `1.13 → 1.14`.  
>     - Validate users are able to perform skip-version upgrade of Consul. For example upgrading from `1.13 → 1.15`.   
> 
> - The upgrade tests also aims to highlight errors that may occur as users attempt to upgrade their current version to a newer version. 

Reviewer: Seems like there are some redundant lines in this section. Also, may want to highlight the goals. Something like:

```suggestion
The goal of upgrade tests is to ensure problem-free upgrades on supported upgrade paths. At any given time, Consul supports the latest minor release, and two older minor releases, e.g. 1.15, 1.14, and 1.13. Upgrades to any higher version are permitted, including skipping a minor version e.g. from 1.13 to 1.15.
```

Here is an example of how the upgrade tests work
1. Create a cluster with a specified number of server and client agents, then enable the feature to be tested. 

> 2. Create some workload in the cluster, e.g. register 2 services: static-server, static-client. 

Reviewer: Considering this was introduced as an example, you can drop the "e.g."s :)

> 2. Create some workload in the cluster, e.g. register 2 services: static-server, static-client. 

Reviewer: I don't think it's super clear here how the static server and static client interact

> 3. Verify connection / disconnection (e.g. deny Action) 

Reviewer: Maybe flesh this out a bit more
```suggestion
3. Configure Consul intention to deny connection between static client and server. Ensure that a connection cannot be made.
```

> 4. Upgrade consul cluster and restart the envoy sidecars (we restart envoy sidecar to ensure the upgraded consul binary can read the state from the previous version and generate the correct envoy configurations) 

Reviewer: Stylistically, I try to stick to using Title Case when referring to products, and `monospaced lower case` when referring to specific technical strings, like image names.

```suggestion
4. Upgrade Consul cluster and restart the Envoy sidecars (we restart Envoy sidecar to ensure the upgraded Consul binary can read the state from the previous version and generate the correct Envoy configurations) 
```

5. Verify connection / disconnection (e.g., deny Action)  

**Note** that all consul agents and user workloads such as application services, mesh-gateway are running in docker containers.

> The tests focus on Consul upgrade lifecycle and emit logs to stdout which developers can silence. Please reach out to the CTIA Team for any questions or help getting started.

Reviewer: I'm kind of struggling to relate these clauses?

> These tests focus on Consul upgrade lifecycle.

Probably redundant with the intro paragraph?

> [Containers] emit logs to stdout which developers can silence [with `--follow-logs=false`].

Kind of a detail for later IMO

> Please reach out to the CTIA Team for any questions or help getting started.

Is it ok to put internal team names in public-facing docs? I honestly am not sure.

## Getting Started
### Prerequisites
If you wish to run or add new test cases, the following are required:
- install [Go](https://go.dev/) (the version should match that of our CI config's Go image).
- install [`golangci-lint`](https://golangci-lint.run/usage/install/) 
- install [`Makefile`](https://www.gnu.org/software/make/manual/make.html)
- [`Docker`](https://docs.docker.com/get-docker/) required to run tests locally  

### Running Upgrade integration tests
In CI, the tests are executed against all test cases defined under test/integration/consul-container/test/upgrade. To run a selected test case locally, e.g., testing upgrade from 1.15 to local changed version, we can use the following command:
```sh
make dev-docker
cd ./test/integration/consul-container
go test -v -timeout 30m -run ^TestACL_Upgrade_Node_Token$ ./.../upgrade/ --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false
```

The command above runs a single test. To run the entire upgrade test suite, the following command can be used.
```sh

> go test -v -timeout 30m ./.../upgrade --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false

Reviewer: Could drop the flags with default values. 
```suggestion
go test -v -timeout 30m ./.../upgrade -follow-log=false
```

```
Below are the supported CLI options

> | Tags / Flags      | Default value | Description |

Reviewer: Not sure what you mean here by `Tags`?

| -----------         | ----------- | ----------- |

> | --latest-image      | consul    | Refers to the docker image name to be deployed before upgrade
> | --latest-version    | latest      | Refers to the version or tag of latest image to be deployed before upgrade
> | --target-image      | consul    | Refers to the image name that will be upgrade to
> | --target-version    | local     | Refers to the version of target image to upgrade to

Reviewer: Some corrections for clarity
```suggestion
| --latest-image      | `consul` in OSS, `hashicorp/consulenterprise` in ENT    | Name of the Docker image to deploy initially.
| --latest-version    | latest      | Tag of the Docker image to deploy initially.
| --target-image      | `consul` in OSS, `hashicorp/consulenterprise` in ENT    | Name of the Docker image to upgrade to.
| --target-version    | local     | Tag of the Docker image to upgrade to. `local` is the tag built by `make dev-docker` above.
```

> | -follow-log         |  false    | This optional flag emit event logs

Reviewer: I thought `--follow-log` is true by default?
```suggestion
| -follow-log         |  true    | Emit all container logs.
```

## Adding a new upgrade integration test  
Upgrade integration tests are defined in the [test/integration/consul-container/test/upgrade](/test/integration/consul-container/test/upgrade) subdirectory and new upgrade integration tests should always be added to this location. The test framework uses
[functional table-driven tests in Go](https://yourbasic.org/golang/table-driven-unit-test/) and using function types to modify the base value for each test case.

> For tests with multiple test cases, it should always start by invoking 

Reviewer: If we are going to standardize on this form (a question for another thread maybe), we should adapt the others to follow suit.

```go
	type testcase struct {
		name string
		create func()
		extraAssertion func()

> 	}

Reviewer: The example below for `l7_traffic_management` has different method signatures than these :P
```
		create func(*libcluster.Cluster, libservice.Service) (libservice.Service, func(), func(), error)
		extraAssertion func(libservice.Service)
```

### How it works
![Upgrade Tests Workflow](util/upgrade_tests_workflow.png?raw=true)

A Consul cluster is deployed, then a static-server, static-client and envoy sidecars are created in the cluster. An API request is made to the static-client to validate that it is ready. 

Then we validate traffic between the static-server and static-client envoy sidecar. After validation, we take a snapshot and upgrade the Consul cluster to the `target-version` and restart the sidecars. Re-validate the client, server and sidecars to ensure the data snapshotted from the pervious version can be accessed in the latest version. 

### Errors Test Cases
There are some caveats for special error handling of versions prior to `1.14`.   
Upgrade tests for features such peering, had API changes that returns an error if attempt to upgrade, and should be accounted for in upgrade tests. If running upgrade tests for any version before `1.14`, the following lines of code needs to be added to skip test or it will not pass.  

```go
		fromVersion, err := version.NewVersion(utils.LatestVersion)
		require.NoError(t, err)
		if fromVersion.LessThan(utils.Version_1_14) {

> 			continue

Reviewer: I like to use e,g, `t.Skipf("version %s (< 1.14) doesn't support <blah>", utils.LatestVersion)`

		}
```
See example [here](https://github.com/hashicorp/consul-enterprise/blob/005a0a92c5f39804cef4ad5c4cd6fd3334b95aa2/test/integration/consul-container/test/upgrade/peering_control_plane_mgw_test.go#L92-L96)

To write tests for bugs found during upgrades, see example on how to add a testcase for those scenarios [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/fullstopupgrade_test.go). 

> 

Reviewer: We should have a section about how we do port forwarding. The stuff around the `pod` containers is especially out of the ordinary.

# Upgrade Integration Tests

The upgrade integration test is responsible for ensuring that users are able to upgrade between consul versions without experiencing difficulties. The upgrade tests does the following:
- Ensure that customers are able to upgrade to a new Consul version and new and existing features work as expected. 

> - Test for backwards compatibility. i.e support for upgrading directly from Consul `1.13` to `1.15`, without upgrading to `1.14`

Reviewer 2: Please explicitly list we upgrade test n-2 and n-1 versions and support skip-version upgrade.

>     - Highlights errors that will occur if attempting to skip a version during upgrades. Example the API changed between version 1.13 and 1.15 so direct upgrades between these version is not supported. 

Reviewer 2: I don't quite understand this point, could you please elaborate a bit? Thanks.

Author: I am describing here the importance of the upgrade tests. The upgrade tests will highlight errors that may occur as customers attempt to upgrade their current version to the latest version. I was then trying to give an example of the API data structure change that prevents users from upgrading from 1.12.3 -> 1.13.

> 

Reviewer 2: In this introductory section, we could briefly explains how the end to end upgrade test works, like follows

```markdown
Following is the main steps of an upgrade test case:

1. Create a cluster with a specified number of server and client agents, and enable the feature to be tested.
2. Create some workloads in the cluster, e.g., register 2 services: static-server, static-client.
3. Verify connection / disconnection (e.g., deny Action)
4. Upgrade consul cluster and restart the envoy sidecars (we restart envoy sidecar to ensure the upgraded consul binary can read the state from the previous version and generate the correct envoy configurations)

5. Verify connection / disconnection (e.g., deny Action)

```

> The tests focus on Consul upgrade lifecycle and emits logs to stdout which developers can be silenced. Please reach out to the [CTIA Team](https://hashicorp.slack.com/app_redirect?channel=team-consul-test-infra) for any questions or help getting started.

Reviewer 2: - nit: emits -> emit

## Getting Started

> We use [`golangci-lint`](https://golangci-lint.run/usage/install/), [`Makefile`](https://www.gnu.org/software/make/manual/make.html) and [`Docker`](https://docs.docker.com/get-docker/) in this repository. It's generally simpler to install the binaries rather than build from source. 

Reviewer 2: Add `### Prerequisites` like https://github.com/hashicorp/consul/blob/main/.github/CONTRIBUTING.md#prerequisites

### Running Upgrade integration tests

> To run integration tests locally,here are the commands needed:

Reviewer 2: nit: space between `,` and `here`.

```sh
make dev-docker
cd ./test/integration/consul-container
go test -v -timeout 30m -run ^TestACL_Upgrade_Node_Token$ ./.../upgrade/ --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false
```

> 

Reviewer 2: Please add some descriptions about the cli parameters, e.g.,

```
* --latest-version
* --target-version: (explains where local comes from)
* -follow-log
```

Upgrade integration tests are defined in the [test/integration/consul-container/test/upgrade](/test/integration/consul-container/test/upgrade) subdirectory and new upgrade integration tests should always be added to this location. The test framework uses
[functional table-driven tests in Go](https://yourbasic.org/golang/table-driven-unit-test/) and using function types to modify the base value for each test case.

For integration tests with multiple test cases, it should always start by invoking 
```go
	type testcase struct {
		name string
		create func()
		extraAssertion func()
	}
```
see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/l7_traffic_management/resolver_default_subset_test.go). For upgrade tests with a single test case, they can be written like
```go
	run := func(t *testing.T, oldVersion, targetVersion string) {
        // insert test
    }
	t.Run(fmt.Sprintf("Upgrade from %s to %s", utils.LatestVersion, utils.TargetVersion),
		func(t *testing.T) {
			run(t, utils.LatestVersion, utils.TargetVersion)
		})
```
see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/acl_node_test.go)

> ### How it works

Reviewer 2: I think visualization like the following will help developers establish a mental picture of the test scenario. Could you add a paragraph on how the test scenario is built.

<img width="549" alt="Screen Shot 2023-03-13 at 10 40 11 AM" src="https://user-images.githubusercontent.com/463631/224734653-a2f7f12b-afb9-4507-856b-32d6c9c6dc51.png">
rios.

- This tests aims to ensure a high degree of interoperability between versions of Consul by supporting `n-2` and `n-1` upgrades, ensuring that upgrading to newer versions does not degrade or break functionalities.
    - Validate users are able to perform a direct upgrade of Consul. For example upgrading from `1.13 → 1.14`.  
    - Validate users are able to perform skip-version upgrade of Consul. For example upgrading from `1.13 → 1.15`.   

- The upgrade tests also aims to highlight errors that may occur as users attempt to upgrade their current version to a newer version. 

Here is an example of how the upgrade tests work
1. Create a cluster with a specified number of server and client agents, then enable the feature to be tested. 
2. Create some workload in the cluster, e.g. register 2 services: static-server, static-client. 
3. Verify connection / disconnection (e.g. deny Action) 
4. Upgrade consul cluster and restart the envoy sidecars (we restart envoy sidecar to ensure the upgraded consul binary can read the state from the previous version and generate the correct envoy configurations) 
5. Verify connection / disconnection (e.g., deny Action)

> The tests focus on Consul upgrade lifecycle and emit logs to stdout which developers can silence. Please reach out to the [CTIA Team](https://hashicorp.slack.com/app_redirect?channel=team-consul-test-infra) for any questions or help getting started.

Reviewer 2: I don't think we need to expose the internal team name and slack link in public repo, so reach out to Consul team should be enough.

## Getting Started
### Prerequisites
If you wish to run or add new test cases, the following are required:
- install [Go](https://go.dev/) (the version should match that of our CI config's Go image).
- install [`golangci-lint`](https://golangci-lint.run/usage/install/) 
- install [`Makefile`](https://www.gnu.org/software/make/manual/make.html)
- [`Docker`](https://docs.docker.com/get-docker/) required to run tests locally  

### Running Upgrade integration tests

> To run upgrade integration tests locally, here are the commands needed:

Reviewer 2: In CI, the tests are executed against all test cases defined under `test/integration/consul-container/test/upgrade`.  To run a selected test case locally, e.g., testing upgrade from 1.15 to local changed version, we can use the following command:

```sh
make dev-docker
cd ./test/integration/consul-container
go test -v -timeout 30m -run ^TestACL_Upgrade_Node_Token$ ./.../upgrade/ --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false
```

The command above runs a single test. To run the entire upgrade test suite, the following command can be used.
```sh
go test -v -timeout 30m ./.../upgrade --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false
```

> 

Reviewer 2: Supported cli options are

[functional table-driven tests in Go](https://yourbasic.org/golang/table-driven-unit-test/) and using function types to modify the base value for each test case.

For tests with multiple test cases, it should always start by invoking 
```go
	type testcase struct {
		name string
		create func()
		extraAssertion func()
	}
```
see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/l7_traffic_management/resolver_default_subset_test.go). For upgrade tests with a single test case, they can be written like
```go
	run := func(t *testing.T, oldVersion, targetVersion string) {
        // insert test
    }
	t.Run(fmt.Sprintf("Upgrade from %s to %s", utils.LatestVersion, utils.TargetVersion),
		func(t *testing.T) {
			run(t, utils.LatestVersion, utils.TargetVersion)
		})
```
see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/acl_node_test.go)

### How it works
![Upgrade Tests Workflow](util/upgrade_tests_workflow.png?raw=true)

> 

Reviewer 2: I don't think this 2 paragraph reveals the details of how it works. Let's think about how to rephrase them.

Author: Okay, I'm open to ideas on how to phrase it.
